FEATURE: Add user custom fields to user directory

markvanlan opened the pull request:
![image](https://user-images.githubusercontent.com/16214023/120380450-cd720380-c2e6-11eb-9fff-a06704054ef4.png)


![image](https://user-images.githubusercontent.com/16214023/120380396-bb906080-c2e6-11eb-9c79-138c0c9ce7d3.png)


![image](https://user-images.githubusercontent.com/16214023/120380355-ae737180-c2e6-11eb-9810-83436e6cc8e8.png)

## Mobile 

![image](https://user-images.githubusercontent.com/16214023/120523693-cf49ce80-c39b-11eb-8abe-f0b4fcaaf860.png)

markvanlan commented on app/assets/javascripts/discourse/app/controllers/edit-user-directory-columns.js:
```diff
@@ -0,0 +1,98 @@
+    let resetColumns = this.columns;
+    resetColumns
+      .sort((a, b) =>
+        (a.automatic_position || a.user_field.position + 1000) >
```
This sorts all the columns with `automatic_position` to the top, then orders the user_field columns properly based on their position. If we ever got over 1000 automatic columns by default we have lost our way and are out of business

markvanlan commented on app/assets/javascripts/discourse/app/controllers/users.js:
```diff
@@ -9,23 +10,30 @@ export default Controller.extend({
   queryParams: ["period", "order", "asc", "name", "group", "exclude_usernames"],
   period: "weekly",
-  order: "likes_received",
+  order: "",
```
We can't have a default value for the query param, because the columns are dynamic. Have to wait until the columns are loaded, then use the first column as the default order

eviltrout commented on app/assets/javascripts/discourse/app/components/directory-table.js:
```diff
@@ -0,0 +1,19 @@
+export default Component.extend({
+  classNames: ["directory-table-container"],
+
+  setActiveHeader(id) {
```
I like the pattern of the child calling up to this action, but I think `id` is inefficient. In the child you could pass `this.element`, then you wouldn't need to `getElementById`. You might not need an `id` at all in the child.

markvanlan replied:
I swear `this.element` was null, thats why I did this mess. I tried it again and now I can access it without directly accessing the dom. Fantastic!

eviltrout commented on app/assets/javascripts/discourse/app/components/directory-table.js:
```diff
@@ -0,0 +1,19 @@
+  setActiveHeader(id) {
+    // After render, scroll table left to ensure the order by column is visible
+    const headerEl = document.getElementById(id);
+    const tableContainer = document.querySelector(".directory-table-container");
```
This could be `this.element` no?

markvanlan replied:
same as above, idk what was going on.

eviltrout commented on app/assets/javascripts/discourse/app/components/table-header-toggle.js:
```diff
@@ -10,15 +8,8 @@ export default Component.extend({
-    return I18n.t(labelKey + "_long", { defaultValue: I18n.t(labelKey) });
-  },
+  translated: false,
+  onActiveRender: null,
```
I think `headerDidRender` might be more descriptive here.

markvanlan replied:
That's not really true though. It's only on render, when the header is "active", meaning the table is being ordered by that header.

eviltrout replied:
Oh I see, that makes sense. Carry on!

eviltrout commented on app/assets/javascripts/discourse/app/helpers/application.js:
```diff
@@ -50,3 +50,7 @@ registerUnbound("number", (orig, params) => {
   return htmlSafe(result);
 });
+
+registerUnbound("value-for-key", (obj = {}, key) => {
```
I sent you a message about this, but I think this can be removed and we can use the `{{get}}` helper instead.

eviltrout commented on app/assets/javascripts/discourse/app/helpers/directory-item-user-field-value.js:
```diff
@@ -0,0 +1,16 @@
+        ? args.item.user.user_fields[args.column.user_field_id]
+        : null;
+    const content = value || "-";
+    return htmlSafe(`<span class='user-field-value'>${content}</span>`);
```
I wonder if these helpers might be better as simple components. I'm on the fence. I suspect we can try it for now.

eviltrout commented on app/assets/javascripts/discourse/app/routes/users.js:
```diff
@@ -36,11 +37,16 @@ export default DiscourseRoute.extend({
-    return params;
+    const columns = PreloadStore.get("directoryColumns");
+    if (!params.order) {
+      params.order = columns[0].name;
```
Trivial, but this if block could be one line: `params.order = params.order || columns[0].name;`

eviltrout commented on app/assets/javascripts/discourse/app/templates/modal/edit-user-directory-columns.hbs:
```diff
@@ -0,0 +1,48 @@
+      {{#each columns as |column|}}
+        <div class="edit-directory-column">
+          <div class="left-content">
+            <label class="column-name">
```
This is repeated almost line for line in an earlier component. Maybe make a component to DRY it up?

markvanlan replied:
I think you're talking about this: 
https://github.com/discourse/discourse/blob/300f4c910f9495d6fe1664f95239b376c7eee42e/app/assets/javascripts/discourse/app/templates/modal/edit-user-directory-columns.hbs#L11-L14

Which looks visually the same as this 
https://github.com/discourse/discourse/blob/300f4c910f9495d6fe1664f95239b376c7eee42e/app/assets/javascripts/discourse/app/templates/components/directory-item.hbs#L4-L8

But use all different helpers

eviltrout replied:
My bad! Sorry.

eviltrout commented on app/serializers/directory_item_serializer.rb:
```diff
@@ -4,6 +4,17 @@ class DirectoryItemSerializer < ApplicationSerializer
+    attributes :user_fields
+
+    def user_fields
+      # allowed_keys = scope.allowed_user_field_ids(object)
```
Looks like this comment was left in by mistake.

eviltrout commented on db/migrate/20210527131318_create_directory_columns.rb:
```diff
@@ -0,0 +1,39 @@
+      t.integer :position, null: false
+      t.datetime :created_at, default: -> { 'CURRENT_TIMESTAMP' }
+    end
+
```
I am always suspicious of tables without indexes. I know this table is likely to be small but we should add indexes anyway. I recommend `enabled, position`, but if it's ever filtered by `user_field_id` that might be a good idea too.

eviltrout reviewed (commented):
It's looking good! Can't wait to see it in action.

markvanlan commented:
@eviltrout I fixed the things you mentioned. One other thing I realized is that I would have allowed disabling all columns, so I added a check in the controller.

eviltrout reviewed (approved):
Looks good! Let's merge maybe on Monday so you are around to debug any issues without ruining your weekend.
